tests/src/core/testqgsgeometry.cpp

```diff
@@ -17887,40 +17888,29 @@ void TestQgsGeometry::poleOfInaccessibility()
+  QTest::newRow( "polygon with extra tail (a part of the ring does not form any area)" ) << QStringLiteral( "POLYGON((0 0, 1 0, 1 1, 0 1, 0 0, -1 0, 0 0))" ) << QStringLiteral( "GeometryCollection (Polygon ((0 1, 1 1, 1 0, 0 0, 0 1)),LineString (0 0, -1 0))" );
+  QTest::newRow( "collection with invalid geometries" ) << QStringLiteral( "GEOMETRYCOLLECTION(LINESTRING(0 0, 0 0), POLYGON((0 0, 10 10, 10 0, 0 10, 0 0)), LINESTRING(10 0, 10 10))" ) << QStringLiteral( "GeometryCollection (Point (0 0),MultiPolygon (((0 0, 0 10, 5 5, 0 0)),((10 0, 5 5, 10 10, 10 0))),LineString (10 0, 10 10))" );
+  QTest::newRow( "null line (#18077)" ) << QStringLiteral( "MultiLineString ((356984.0625 6300089, 356984.0625 6300089))" ) << QStringLiteral( "Point (356984.0625 6300089)" );
+}
+
 void TestQgsGeometry::makeValid()
 {
-  typedef QPair<QString, QString> InputAndExpectedWktPair;
-  QList<InputAndExpectedWktPair> geoms;
-  // dimension collapse
-  geoms << qMakePair( QStringLiteral( "LINESTRING(0 0)" ),
-                      QStringLiteral( "POINT(0 0)" ) );
-  // unclosed ring
-  geoms << qMakePair( QStringLiteral( "POLYGON((10 22,10 32,20 32,20 22))" ),
-                      QStringLiteral( "POLYGON((10 22,10 32,20 32,20 22,10 22))" ) );
-  // butterfly polygon (self-intersecting ring)
-  geoms << qMakePair( QStringLiteral( "POLYGON((0 0, 10 10, 10 0, 0 10, 0 0))" ),
-                      QStringLiteral( "MULTIPOLYGON(((5 5, 0 0, 0 10, 5 5)),((5 5, 10 10, 10 0, 5 5)))" ) );
-  // polygon with extra tail (a part of the ring does not form any area)
-  geoms << qMakePair( QStringLiteral( "POLYGON((0 0, 1 0, 1 1, 0 1, 0 0, -1 0, 0 0))" ),
-                      QStringLiteral( "GEOMETRYCOLLECTION(POLYGON((0 0, 0 1, 1 1, 1 0, 0 0)), LINESTRING(0 0, -1 0))" ) );
-  // collection with invalid geometries
-  geoms << qMakePair( QStringLiteral( "GEOMETRYCOLLECTION(LINESTRING(0 0, 0 0), POLYGON((0 0, 10 10, 10 0, 0 10, 0 0)), LINESTRING(10 0, 10 10))" ),
-                      QStringLiteral( "GEOMETRYCOLLECTION(POINT(0 0), MULTIPOLYGON(((5 5, 0 0, 0 10, 5 5)),((5 5, 10 10, 10 0, 5 5))), LINESTRING(10 0, 10 10))" ) );
-  // null line (#18077)
-  geoms << qMakePair( QStringLiteral( "MultiLineString ((356984.0625 6300089, 356984.0625 6300089))" ),
-                      QStringLiteral( "MultiPoint ((356984.0625 6300089))" ) );
-
-  for ( const InputAndExpectedWktPair &pair : geoms )
-  {
-    QgsGeometry gInput = QgsGeometry::fromWkt( pair.first );
-    QgsGeometry gExp = QgsGeometry::fromWkt( pair.second );
-    QVERIFY( !gInput.isNull() );
-    QVERIFY( !gExp.isNull() );
-
-    QgsGeometry gValid = gInput.makeValid();
-    QVERIFY( gValid.isGeosValid() );
-    QVERIFY( gValid.isGeosEqual( gExp ) );
-  }
+  QFETCH( QString, input );
+  QFETCH( QString, expected );
+
+  QgsGeometry gInput = QgsGeometry::fromWkt( input );
+  QVERIFY( !gInput.isNull() );
+
+  QgsGeometry gValid = gInput.makeValid();
+  QCOMPARE( gValid.asWkt(), expected );
```

Comment:
GEOSNormalize_r() should probably be used to compare actual vs expected results to be independent on the vertex order

Reply:
Ok... I'll need to port that to QGIS first so that we can handle M / curved geometries!

Reply:
Done!